### PR TITLE
Change XML Unmarshaling of the alias names of devices of a domain to …

### PR DESCRIPTION
**Pull request by yuhaohaoyu**
…reflect the corresponding Marshaling.

Signed-off-by: Hao Yu <yuh@us.ibm.com>

**What this PR does / why we need it**:

In KubeVirt, the virt-launcher directly communicates to the libvirt, where an virtual machine instance object's domain specification is passed between the two parties in XML. There is a customized treatment for the conversion between the KubeVirt `Domain` interface and Libvirt's XML format for `user defined` alias names of various devices. 

Following some earlier feedbacks from the reviewers/collaborators, this PR is set to meet the following design to pertain the consistency for xml-marshaling/unmarshaling of the `user defined` alias names.
- Make the `Name` attribute of the type `Alias` in `pkg/virt-launcher/virtwrap/api/schema.go` private, add a `userDefined` boolean into the type.
- Add an explicit constructors: `func NewUserDefinedAlias(name string) Alias` and make use of it in `pkg/virt-launcher/virtwrap/converter/converter.go`
- In the Marshal and Unmarshal overloads for XML, properly determine whether there should be the prefix, trim the prefix and set userDefined properly and vice-versa.

These changes affect mainly the `converter` with many unit tests under `pkg/virt-launcher` and `pkg/virt-handler`. The changes are in 4 aspects:
- Define the exported constructor, getter, xml-marshal/unmarshal overloads (`virt-launcher/virtwrap/api/schema.go`).
- Modify how the `Alias` in KubeVirt `Domain` objects are constructed (aka to use the `NewUserDefinedAlias()` constructor).
- Modify how the `Name` attribute of the type `Alias` is referenced (read reference via a getter: `func GetName() string`).
- Modify some XML literals in some unit tests, e.g. change `<alias name="ua-mydisk"></alias>` to `<alias name="ua-mydisk" userdefined="true"></alias>`, etc. 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

This may have a positive effect to the downstream bug: https://bugzilla.redhat.com/show_bug.cgi?id=1911118

**Release note**:

```release-note
"NONE"
```

### CODE REVIEW

**Review by ashleyschuett — COMMENTED** [review]
can we add a unit test case for this?

File: pkg/virt-launcher/virtwrap/api/schema.go [hunk 1]
```diff
@@ -683,7 +684,7 @@ func (alias *Alias) UnmarshalXML(d *xml.Decoder, start xml.StartElement) error {
 		return err
 	}
 	*alias = Alias(userAlias)
-	alias.Name = alias.Name[len(UserAliasPrefix):]
+	alias.Name = strings.TrimPrefix(alias.Name, UserAliasPrefix)
```
Comment (vladikr) [inline pkg/virt-launcher/virtwrap/api/schema.go:687]: I think that `TrimPrefix` kind of does the same, no? 
https://golang.org/src/strings/strings.go?s=22858:22898#L892

**Review by vladikr — COMMENTED** [review]
I think the description of this PR should be updated. 

The only difference between the changes is that so far we always assumed that `UserAliasPrefix` being added to any device alias. I don't know if this is related to the d/s BZ

File: pkg/virt-launcher/virtwrap/api/schema.go (hunk 1 above)
Comment (yuhaohaoyu) [reply pkg/virt-launcher/virtwrap/api/schema.go:687]: Following the same link, https://golang.org/src/strings/strings.go?s=22858:22898#L892, `TrimPrefix` does: `// If s doesn't start with prefix, s is returned unchanged.` 

The reason for this PR was that we have seen that some devices' alias names got evolved into things like `ua-let0` (from `tablet0`), `ua-tio` (from `virtio`) in downstream tests involving Windows VM migration and shutdown. So the speculated lifecycle of these alias names, `tablet0` as an example, would likely be: 
      ```tablet0 --UnmarshalXML--> let0 --MarshalXML--> ua-let0``` 

For the said bug scenario, we could not quite figure out why the "unmarshalled" alias name, e.g. `tablet0`, was originated. But this fix will prevent such short "unmarshalled" alias name got trimmed shorter, leading to multiple devices having the same alias name.

Comment (vladikr) [reply pkg/virt-launcher/virtwrap/api/schema.go:687]: Sure. I'm not against this change :) I'm just asking to update the description.

**Comment by yuhaohaoyu** [comment]
> can we add a unit test case for this?

Added a unit test scenario to `Unmarshal` device alias names starting with `ua-` and those not starting with `ua-`.

File: pkg/virt-launcher/virtwrap/api/schema_test.go [hunk 2]
```diff
@@ -189,6 +191,7 @@ var _ = Describe("Schema", func() {
 		`<memballoon model="virtio">
       <stats period="10"></stats>
     </memballoon>`)
+	exampleXMLwithDeviceWithoutUserAliasPrefix = strings.Replace(exampleXML, "ua-tablet0", "tablet0", 1)
```
Comment (rmohr) [inline pkg/virt-launcher/virtwrap/api/schema_test.go:194]: Looks like you can keep this completely in the test. I don't think you need to have a global function or do something in BeforeEach. Just prepare it in `It`.

File: pkg/virt-launcher/virtwrap/api/schema.go (hunk 1 above)
Comment (rmohr) [reply pkg/virt-launcher/virtwrap/api/schema.go:687]: I would suggest that we fix it completely. The current change only fixes it for the first unmarshalling. Subsequent marshallings will then serialize it then without the `ua-` prefix, no matter if it had one before or not.

Can we do the following:

 1) the current `UserAlias` struct makes no sense at all, it is only used in the marshal and unmarshal functions and can as well be removed completely.
 2) If we really want to hide this `ua-` prefix handling correclty, I think we should make the `Name`  attribute private in the struct, add a `userDefined` boolean to the struct and then:
 3) Add an explicit constructors: `func UserDefinedAlias(name string) Alias` and make use of it in converter.go
 4) In the Marshal and Unmarshal overloads for XML, we can then properly detect the prefix , trim the prefix and set `userDefined` properly and vice-versa.

**Review by rmohr — CHANGES_REQUESTED** [review]
That's a nice discovery of an ugly small issue. This fix is definitely an improvement. I make some suggestions to fix the issue for all scenarios.

File: pkg/virt-launcher/virtwrap/api/schema.go (hunk 1 above)
Comment (yuhaohaoyu) [reply pkg/virt-launcher/virtwrap/api/schema.go:687]: @rmohr, the key of your layout appears to rely on the knowledge that uses of the explicit constructor for `UserAlias` object are contained only in `converter.go`. I will go ahead and code in this layout.

Comment (rmohr) [reply pkg/virt-launcher/virtwrap/api/schema.go:687]: > @rmohr, the key of your layout appears to rely on the knowledge that uses of the explicit constructor for `UserAlias` object are contained only in `converter.go`. I will go ahead and code in this layout.

That's exactly it. Perfect :+1:

File: pkg/virt-launcher/virtwrap/api/schema.go [hunk 3]
```diff
@@ -661,29 +662,50 @@ type InterfaceTarget struct {
...
+
+func (alias Alias) GetName() string {
+	return alias.name
+}
+
 func (alias Alias) MarshalXML(e *xml.Encoder, start xml.StartElement) error {
-	userAlias := UserAlias(alias)
-	userAlias.Name = UserAliasPrefix + userAlias.Name
-	return e.EncodeElement(userAlias, start)
+	userAliasXml := userAliasXML{Name: alias.name, UserDefined: alias.userDefined}
```
Comment (rmohr) [inline pkg/virt-launcher/virtwrap/api/schema.go:692]: The public visible part looks fine but you have to change `MarshalXML` and `UnmarshalXML`: We have to provide the same content like before the libvirt. We can't just invent a new boolean field in the xml.

Consider something like this for marshalling:

```go
func (alias Alias) MarshalXML(e *xml.Encoder, start xml.StartElement) error {
        name := alias.name
        if alias.userDefined {
          name = UserAliasPrefix + name
        }
	start.Attr = []xml.Attr{xml.Attr{Name: xml.Name{Local: "name"}, Value: name}}
	e.EncodeToken(start)
	e.EncodeToken(xml.EndElement{Name: start.Name})
}
```

And vice-versa:

```go
func (alias *Alias) UnmarshalXML(d *xml.Decoder, start xml.StartElement) error {
	var name string
	for _ ,attr := range start.Attr {
		if attr.Name.Local == "name" {
			name = attr.Value
			break
		}
	}
	if strings.HasPrefix(name, UserAliasPrefix) {
		alias.UserDefined = true
	}
	alias.Name = strings.TrimPrefix(name)
	return nil
}
```

**Review by rmohr — CHANGES_REQUESTED** [review]
Proposed some serialization and deserialization methods which you can use. We must not invent new fields on the domain xml.

File: hack/bazel-test.sh [hunk 4]
```diff
@@ -5,4 +5,6 @@ source hack/config.sh
 
 bazel test \
     --config=${ARCHITECTURE} \
-    --test_output=errors -- //staging/src/kubevirt.io/client-go/... //pkg/... //cmd/... //tests/framework/...
+    --test_output=errors -- //pkg/virt-handler/... //pkg/virt-launcher/...
+
+#    --test_output=errors -- //staging/src/kubevirt.io/client-go/... //pkg/... //cmd/... //tests/framework/...
```
Comment (rmohr) [inline hack/bazel-test.sh:10]: That seems like it was added by accident.

**Comment by rmohr** [comment]
There is something wrong with the rebate.

File: hack/bazel-test.sh (hunk 4 above)
Comment (yuhaohaoyu) [reply hack/bazel-test.sh]: > There is something wrong with the rebate.

@rmohr, I tried quite a few things but could not figure how to fix the rebase issue. Can you please give me some guide on this? I think I can do something like 'git reset --hard' to some commit and I can re-apply the changes manually from there.

Comment (rmohr) [reply hack/bazel-test.sh]: git fetch origin master
git rebase -i origin/master

That should first show you that you rebase a single commit, then you can resolve possible merge conflicts.

When done, force push on your remote branch.

**Comment by yuhaohaoyu** [comment]
#4927 
I apologize for failing the conflict merging effort. I have created a new PR carrying the same change-set, without conflicts.

**Comment by rmohr** [comment]
/close

as per this comment:

> #4927
> I apologize for failing the conflict merging effort. I have created a new PR carrying the same change-set, without conflicts.

closing.

**Pull request closed without merging**